src/foam/nanos/auth/AgentUserAuthService.java

```diff
@@ -0,0 +1,126 @@
+import foam.core.X;
+import foam.dao.DAO;
+import foam.nanos.NanoService;
+import foam.nanos.session.Session;
+
+import static foam.mlang.MLang.EQ;
+import static foam.mlang.MLang.AND;
+
+public class AgentUserAuthService
+  extends    ContextAwareSupport
+  implements AgentAuthService, NanoService
+{
+  protected DAO userDAO_;
+  protected DAO groupDAO_;
+  protected DAO sessionDAO_;
+  protected DAO agentJunctionDAO_;
+
+  public AgentUserAuthService(X x) {
+    setX(x);
+  }
+
+  @Override
+  public void start() {
+    userDAO_     = (DAO) getX().get("localUserDAO");
+    groupDAO_    = (DAO) getX().get("groupDAO");
+    sessionDAO_  = (DAO) getX().get("sessionDAO");
+    agentJunctionDAO_  = (DAO) getX().get("agentJunctionDAO");
+  }
+
+  /* 
+    Sets the currently logged in user as an "agent" within the context &
+    sets "user" in the context to the passed in user. This allows users to
+    act on behalf of others while retaining information on the user.
+  */
+  public User actAs(X x, User superUser) throws AuthenticationException {
+    User agent = (User) x.get("user");
+    User user = (User) userDAO_.inX(x).find(superUser.getId());
+
+    // Check for current context user
+    if ( agent == null ) {
+      throw new AuthenticationException();
+    }
+
+    if ( user == null ) {
+      throw new AuthorizationException("Super user doesn't exist.");
+    }
+
+    Group group = (Group) groupDAO_.inX(x).find(user.getGroup());
+
+    if ( group == null || ! group.getEnabled() ) {
+      throw new AuthorizationException("Super user must exist within a group.");
+    }
+
+    /*
+      Finds the UserUserJunction object to see if user can act as the passed in user.
+      Source (agent) users are permitted to act as target (superUser) users, not vice versa.
+    */
+    UserUserJunction permissionJunction = (UserUserJunction) agentJunctionDAO_.inX(x).find(AND(
+      EQ(UserUserJunction.SOURCE_ID, agent.getId()),
+      EQ(UserUserJunction.TARGET_ID, user.getId())
+    ));
+
+    if ( permissionJunction == null ) {
+      throw new AuthorizationException("You don't have access to act as the requested super user.");
+    }
+
+    // Junction object contains a group which has a unique set of permissions specific to the relationship.
+    Group actingWithinGroup = (Group) groupDAO_.inX(x).find(permissionJunction.getGroup());
+
+    if ( actingWithinGroup == null || ! actingWithinGroup.getEnabled() ) {
+      throw new AuthorizationException("No permissions are appended to the super user relationship.");
+    }
+    
+    // Clone user and associate new junction group to user. Clone and freeze both user and agent. 
+    user = (User) user.fclone();
+    user.setGroup(actingWithinGroup.getId());
+    user.freeze();
+
+    agent = (User) agent.fclone();
+    agent.freeze();
+
+    // Set user and agent objects into the session context and place into sessionDAO.
+    Session session = x.get(Session.class);
+    session.setUserId(user.getId());
```

Comment:
Do we want to update the session user id?

Reply:
Sure, considering it's used as a reference to the user within the context. We'd still want it to relate to the new entity we put into the context.

Reply:
I don't think it is used. When the user is needed from the context we grab "user". This is just used to identity the user in the session object for things like only enabling one login per user.

Reply:
It's used in https://github.com/foam-framework/foam2/blob/95b422f4b9928107f5ed810ec1190162a8216ba7/src/foam/nanos/auth/UserAndGroupAuthService.java#L59

If we didn't set the userId in the context it would return the wrong user.

Reply:
Can we modify UserAndGroupAuthService to just pull the User out of the context instead?

Reply:
There are other areas that have use the userId to pull the user from the session. I don't think going around and changing all of those places. Unless you can think of a good reason to do that.

Reply:
Can you provide me with a list of places that do that?

Reply:
https://github.com/foam-framework/foam2/blob/0c9301e7a8e81937b67dc9955e569159977c397e/src/foam/box/SessionServerBox.java#L63

https://github.com/foam-framework/foam2/blob/0c9301e7a8e81937b67dc9955e569159977c397e/src/foam/nanos/auth/Group.js#L178

https://github.com/foam-framework/foam2/blob/0c9301e7a8e81937b67dc9955e569159977c397e/src/foam/nanos/auth/UserAndGroupAuthService.java#L54

https://github.com/foam-framework/foam2/blob/0c9301e7a8e81937b67dc9955e569159977c397e/src/foam/nanos/auth/UserAndGroupAuthService.java#L59

https://github.com/foam-framework/foam2/blob/0c9301e7a8e81937b67dc9955e569159977c397e/src/foam/nanos/auth/UserAndGroupAuthService.java#L223

https://github.com/foam-framework/foam2/blob/0c9301e7a8e81937b67dc9955e569159977c397e/src/foam/nanos/auth/UserAndGroupAuthService.java#L228